std/algebra/emulated/sw_emulated/point.go

```diff
@@ -1255,3 +1247,553 @@ func (c *Curve[B, S]) MultiScalarMul(p []*AffinePoint[B], s []*emulated.Element[
+
+	// i = 0
+	// subtract Q and R if the first bits are 0.
+	// When cfg.CompleteArithmetic is set, we use AddUnified instead of Add.
+	// This means when s=0 then Acc=(0,0) because AddUnified(Q, -Q) = (0,0).
+	tableQ[0] = addFn(tableQ[0], Acc)
+	Acc = c.Select(s1bits[0], Acc, tableQ[0])
+	tableR[0] = addFn(tableR[0], Acc)
+	Acc = c.Select(s2bits[0], Acc, tableR[0])
+
+	if cfg.CompleteArithmetic {
+		Acc = c.Select(c.api.Or(selector1, selector2), tableR[2], Acc)
+	}
+	// we added [3]R at the last iteration so the result should be
+	// 		Acc = [s1]Q + [s2]R + [3]R
+	// 		    = [s1]Q + [s2*s]Q + [3]R
+	// 		    = [s1+s2*s]Q + [3]R
+	// 		    = [0]Q + [3]R
+	// 		    = [3]R
+	c.AssertIsEqual(Acc, tableR[2])
+
+	return &AffinePoint[B]{
+		X: *R[0],
+		Y: *R[1],
+	}
+}
+
+// scalarMulGLVAndFakeGLV computes [s]P and returns it. It doesn't modify P nor s.
+// It implements the "GLV + fake GLV" explained in [ethresear.ch/fake-GLV].
+//
+// ⚠️  The scalar s must be nonzero and the point Q different from (0,0) unless [algopts.WithCompleteArithmetic] is set.
+// (0,0) is not on the curve but we conventionally take it as the
+// neutral/infinity point as per the [EVM].
+//
+// [ethresear.ch/fake-GLV]: https://ethresear.ch/t/fake-glv-you-dont-need-an-efficient-endomorphism-to-implement-glv-like-scalar-multiplication-in-snark-circuits/20394
+// [EVM]: https://ethereum.github.io/yellowpaper/paper.pdf
+func (c *Curve[B, S]) scalarMulGLVAndFakeGLV(P *AffinePoint[B], s *emulated.Element[S], opts ...algopts.AlgebraOption) *AffinePoint[B] {
+	cfg, err := algopts.NewConfig(opts...)
+	if err != nil {
+		panic(err)
+	}
+
+	// handle zero-scalar
+	var selector0 frontend.Variable
+	_s := s
+	if cfg.CompleteArithmetic {
+		selector0 = c.scalarApi.IsZero(s)
+		_s = c.scalarApi.Select(selector0, c.scalarApi.One(), s)
+	}
+
+	// Instead of computing [s]P=Q, we check that Q-[s]P == 0.
+	// Checking Q - [s]P = 0 is equivalent to [v]Q + [-s*v]P = 0 for some nonzero v.
+	//
+	// The GLV curves supported in gnark have j-invariant 0, which means the eigenvalue
+	// of the GLV endomorphism is a primitive cube root of unity.  If we write
+	// v, s and r as Eisenstein integers we can express the check as:
+	//
+	// 			[v1 + λ*v2]Q + [u1 + λ*u2]P = 0
+	// 			[v1]Q + [v2]phi(Q) + [u1]P + [u2]phi(P) = 0
+	//
+	// where (v1 + λ*v2)*(s1 + λ*s2) = u1 + λu2 mod (r1 + λ*r2)
+	// and u1, u2, v1, v2 < r^{1/4} (up to a constant factor).
+	//
+	// This can be done as follows:
+	// 		1. decompose s into s1 + λ*s2 mod r s.t. s1, s2 < sqrt(r) (hinted classical GLV decomposition).
+	// 		2. decompose r into r1 + λ*r2  s.t. r1, r2 < sqrt(r) (hardcoded half-GCD of λ mod r).
+	// 		3. find u1, u2, v1, v2 < c*r^{1/4} s.t. (v1 + λ*v2)*(s1 + λ*s2) = (u1 + λ*u2) mod (r1 + λ*r2).
+	// 		   This can be done through a hinted half-GCD in the number field
+	// 		   K=Q[w]/f(w).  This corresponds to K being the Eisenstein ring of
+	// 		   integers i.e. w is a primitive cube root of unity, f(w)=w^2+w+1=0.
+	//
```

Comment:
- hardcoding `r1` and `r2` s.t. `r1 + λr2=0 mod r` is a bit of an issue in the hint, because then we currently need a `switch` to retrieve the values per-curve and also we need to revisit the code every time we add a new curve. I believe re-computing in the hint is fine for now.